docs/proposals/caching.adoc

```diff
@@ -0,0 +1,77 @@
+= Proposal: Caching
+
+== Purpose
+
+This proposal aims to support caching of dependencies
+(https://github.com/opendevstack/ods-pipeline/issues/147) in order to speed up build times. 
+
+== Background
+
+Currently there is a single PVC per project which is shared by all build pipelines. 
+As a consequence only a single build at a time can run for a project.
+
+The PVC is mounted as the workspace in all build tasks. 
+
+The `ods-start` task wipes the PVC at the beginning of each build so that no data persists between build.
+
+Each tekton task which gets started mounts the workspace PVC which can take quiet noticeable wait times.   
+
+== Solution
+
+In the initial implementation instead of a single PVC per project one PVC per repo will be used. The PVC will be used both as a workspace for the build and also as a cache of data between builds.
+
+A build task run will have the following directories available:
+
+* A repo checkout dir. This is the current workspace directory although I propose to rename it to checkout directory as for some technologies the actual build will not happen here but in a caching directory if available.
+
+With sufficient caching space available: 
+
+* A global cache directory where the parent directory is indicated by parameter `global-cache-parent`. Cleanup will only spare directories with in it so that build tasks must keep stuff in a subdirectory. For example in `$(global-cache-parent)/go-modules/`. 
+
+* A pipeline cache directory as indicated by parameter `pipeline-cache-dir`.
+
+In the initial implementation the parameters will be set to `/.cache/` and `/.cache-p/<pipeline-name>/` if space permits and be empty if there is not sufficient space available.
+
+While initially the caches will be on the same PVC as the checkout PVC, this should not be relied on, for example by using hard links between the checkout dir and the cache area. In the future we may want to keep different PVC for these areas. For example the global cache parent could be one day per project and perhaps even one per project for each build technology such as go, node. +
+Also in a future version the `pipeline-cache-dir` may be on a different PVC as the checkout PVC for example to enable faster cleanup of the pristine checkout dir by simply recreating it or not even having it on a PVC. 
+
+Files in the cache locations would persist between builds unless they are removed by the cleanup mechanism. 
+
+The build scripts are adjusted to take advantage of the cache parameters above. In addition each build task will support a new parameters `global-caching-mb` and `pipeline-caching-mb`. These defaults to 0 to indicate the task is not using caching. Otherwise it specify the minimum required cache space in MB.    
```

Comment:
What is the reasoning behind two separate parameters? Why do we need both?

Reply:
As now mentioned in the comment above one purpose of these parameters is to enable caching by explicitly stating how much cache space the task will require in the respective cache.
For the pipeline cache this will also allow for a good cleanup strategy as the space is not shared with other pipelines.
For the global cache the situation is more complex as it would be shared by all pipelines. So it would grow over time and cleanup might from time to time wipe out the cache (e.g. `rm -rf /.cache/go`) when the disk space on the caching PVC no longer is able to fulfill the demand. However to do a good job the cleaner would actually need to understand which tasks use which global caches which the proposal as is does not make explicit.